### PR TITLE
ENH: improved accuracy and orthonormality of output eigenvectors in lobpcg

### PR DESCRIPTION
**Pull request by lobpcg**
#### Reference issue
Closes gh-16319, gh-16408. gh-15461, gh-15935

#### What does this implement/fix?

- Introduced a free and simple tunable restart to potentially increase the attainable accuracy for edge cases  
- Internal postprocessing in lobpcg that runs one final exact RR giving more accurate and orthonormal eigenvectors
- Output lists of history are now in np array format, to improve performance a bit
- In cases of iteration failure or not reaching the required tolerance in the given number of iterations, the code now outputs the computed iterate with the smallest max norm of the residual and drops the history of subsequent iterations. 
- Some minor code beatification and restructuring, not affecting the iterates. 
- Replace the internal matrix converter into the LinearOperator with a simple function, removing the unnecessary dependence.
- Remove the check for LinearOperator format input and thus allow a simple function handle of a callable object as an input.
- Better consistency in using dtypes for different objects
- Multiple warnings added for the user to easier see how to fix edge cases or violating the assumptions crucial for the algorithm health
- Fool-proof enhancements, attempting to guess and fix user errors in the input, rather than letting the execution to break

### CODE REVIEW

**Comment by WarrenWeckesser** [comment]
https://xkcd.com/1296/ 😁

**Comment by lobpcg** [comment]
The PR is already at a stage of nitpicking, waiting for reviews, please

**Comment by lobpcg** [comment]
refguide_asv_check fail unrelated to this PR

**Comment by lobpcg** [comment]
This PR is ready for review & merge, please.

**Review by tylerjereddy — COMMENTED** [review]
I read through the code changes and some of the comments in the matching issues. If we frame this as mostly "enhancement" it perhaps becomes less critical for the `1.10` branching in 9 days or so, though I believe one of the testing adjustments does prevent a failure on `aarch64`, which might be useful to have in.

- one problem is that we haven't had a code review since this was opened 6 months ago, but at least the CI is passing and you're obviously an expert on this algorithm
- I added a few comments that were focused on things like backwards compatibility and exception handling
- apart from 1 or 2 comments I added, most of the testing changes seemed to be improvements--I don't know that I saw tests added for some of the new error handling scenarios (or how important those tests would be); some new warnings do seem covered I think
- it looked like there was some positive feedback on the changes here from the JAX team at Google here: https://github.com/scipy/scipy/issues/16408#issuecomment-1171165201; but then they noticed a tradeoff here: https://github.com/scipy/scipy/issues/16408#issuecomment-1204652701; am I reading that correctly? can you summarize that situation for non-expert?
- we'd obviously need to squash merge b/c of the 63+ commits

File: scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py [hunk 1]
```diff
@@ -146,45 +160,52 @@ def lobpcg(
...
     -------
     w : ndarray
         Array of ``k`` eigenvalues.
     v : ndarray
         An array of ``k`` eigenvectors.  `v` has the same shape as `X`.
-    lambdas : list of ndarray, optional
+    lambdas : ndarray, optional
         The eigenvalue history, if `retLambdaHistory` is True.
-    rnorms : list of ndarray, optional
+    rnorms : ndarray, optional
```
[the remark this answers is not in the document]
Comment (lobpcg) [reply scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:208]: Of course, there are concerns, even though it's an optional output. However, the performance gain outweighs the incompatibly, I think.

File: scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py [hunk 2]
```diff
@@ -80,13 +74,30 @@ def _applyConstraints(blockVectorV, factYBY, blockVectorBY, blockVectorY):
...
+                      verbosityLevel=0):
+    """in-place B-orthonormalize the given block vector using Cholesky."""
     normalization = blockVectorV.max(axis=0) + np.finfo(blockVectorV.dtype).eps
     blockVectorV = blockVectorV / normalization
     if blockVectorBV is None:
         if B is not None:
-            blockVectorBV = B(blockVectorV)
+            try:
+                blockVectorBV = B(blockVectorV)
+            except Exception as e:
```
[the remark this answers is not in the document]
Comment (lobpcg) [reply scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:86]: The callable(B) case is the main source of uncertainty. I do not know how to make the Exception more specific. Any advice?

File: scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py [hunk 3]
```diff
@@ -366,44 +417,81 @@ def lobpcg(
...
+                    raise ValueError(
+                        f"The shape {A.shape} of the primary matrix\n"
+                        f"defined by a callable object is wrong.\n"
+                    )
+            elif isspmatrix(A):
+                A = A.toarray()
             else:
-                B = np.asarray(B)
+                A = np.asarray(A)
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (lobpcg) [reply scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:434]: callable(A) is the source of uncertainty. I do not know how to make the Exception more specific. Any advice?

File: scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py [hunk 4]
```diff
@@ -640,49 +743,47 @@ def _handle_gramA_gramB_verbosity(gramA, gramB):
...
-            _handle_gramA_gramB_verbosity(gramA, gramB)
+            _handle_gramA_gramB_verbosity(gramA, gramB, verbosityLevel)
 
             try:
                 _lambda, eigBlockVector = eigh(gramA,
                                                gramB,
                                                check_finite=False)
-            except LinAlgError:
+            except LinAlgError as e:
+                # raise ValueError("eigh failed in lobpcg iterations") from e
```
[the remark this answers is not in the document]
Comment (lobpcg) [reply scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:753]: I do not remember my motivation to keep it, maybe unsure which way is better.

File: scipy/sparse/linalg/_eigen/lobpcg/tests/test_lobpcg.py [hunk 5]
```diff
@@ -383,40 +399,66 @@ def test_maxit():
...
-    n = 40
-    m = 4
     # Define the generalized eigenvalue problem Av = cBv
     # where (c, v) is a generalized eigenpair,
     # and where we choose A  and B to be diagonal.
     vals = np.arange(1, n + 1)
 
-    list_sparse_format = ['bsr', 'coo', 'csc', 'csr', 'dia', 'dok', 'lil']
+    # list_sparse_format = ['bsr', 'coo', 'csc', 'csr', 'dia', 'dok', 'lil']
+    list_sparse_format = ['coo']
```
[the remark this answers is not in the document]
Comment (lobpcg) [reply scipy/sparse/linalg/_eigen/lobpcg/tests/test_lobpcg.py:433]: With the full list, it's one of the 10 longest tests in scipy to my recollection. For the purpose of testing the solver itself, one format is good enough. The full list would be helpful to test all sparse formats but is not the focus of this PR. Plese advise, I have no personal preference either way - just have minimized my time waiting to the tests to run.

**Comment by lobpcg** [comment]
@tylerjereddy - thanks for reviewing!

> * I don't know that I saw tests added for some of the new error handling scenarios (or how important those tests would be); some new warnings do seem covered I think

Yes. It's lots of nontrivial work to come up with examples initiating every new error handling code. I have added only those unit tests that are relatively simple to grenerate. It's still an improvement compared to the original code without this error handling.

> * it looked like there was some positive feedback on the changes here from the JAX team at Google here: [#16408 (comment)](https://github.com/scipy/scipy/issues/16408#issuecomment-1171165201); but then they noticed a tradeoff here: [#16408 (comment)](https://github.com/scipy/scipy/issues/16408#issuecomment-1204652701); am I reading that correctly? can you summarize that situation for non-expert?

There is no tradeoff. The reviewer there and I in my tests get different accuracy for one of the examples of this PR, but the original code simply fails on this example to my recollection, so the PR brings an improvement either way. 

One of the main goals of this PR is very simple - without changing the algorithm make the code as fool-proof as possible, e.g., even in the worst-case scenario of immediate failure just returning the unchanged initial approximations with a proper warning rather than an error.  Allowing the user to call function handles instead of LinearOperators is extremely flexible and convenient but opens a whole new can of warms allowing gross abuses of the underlying assumptions on the input matrices making the algorithm to fail in different parts of the code.

**Comment by tylerjereddy** [comment]
cc @melissawm

File: scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py [hunk 6]
```diff
@@ -146,45 +160,52 @@ def lobpcg(
...
+    retLambdaHistory : bool, optional.
         Whether to return eigenvalue history.  Default is False.
-    retResidualNormsHistory : bool, optional
+    retResidualNormsHistory : bool, optional.
         Whether to return history of residual norms.  Default is False.
+    restartControl : int, optional.
+        Iterations restart if the residuals jump up 2**restartControl times
+        compared to the smallest ones recorded in retResidualNormsHistory.
+        The default is ``restartControl=20``, making the restarts rare for
+        backward compatibility.
```
[the remark this answers is not in the document]
Comment (lobpcg) [reply scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:198]: It's a practical observation from my testing

File: scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py [hunk 7]
```diff
@@ -366,44 +417,81 @@ def lobpcg(
...
+                # Reverse order to be compatible with eigs() in 'LM' mode.
+                vals = vals[::-1]
+                vecs = vecs[:, ::-1]
+
+            return vals, vecs
+        except Exception as e:
+            raise Exception(
+                f"Dense eigensolver failed with error\n"
+                f"{e}\n"
+            )
```
[the remark this answers is not in the document]
Comment (lobpcg) [reply scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:474]: Not just callable. Maybe even a matrix input, but the matrices violate assumptions, e.g.  the symmetry, could result in failure here I think. And the assumptions are too expensive to validate even for matrices...

File: scipy/sparse/linalg/_eigen/lobpcg/tests/test_lobpcg.py [hunk 8]
```diff
@@ -371,10 +386,11 @@ def test_random_initial_float32():
 
 
 def test_maxit():
-    """Check lobpcg if maxit=10 runs 10 iterations
+    """Check lobpcg if maxit=maxiter runs maxiter iterations and
     if maxit=None runs 20 iterations (the default)
     by checking the size of the iteration history output, which should
-    be the number of iterations plus 2 (initial and final values).
+    be the number of iterations plus 3 (initial, final, and postprocessing)
+    typically when maxiter is small and the choice of the best is passive.
```
[the remark this answers is not in the document]
Comment (lobpcg) [reply scipy/sparse/linalg/_eigen/lobpcg/tests/test_lobpcg.py:393]: Great point. Done.

**Review by melissawm — APPROVED** [review]
I'm not super familiar with the algorithm but this looks like an improvement to me. I'm wondering if there are any changes in performance, but also not sure how easy those would be to check. Overall I'm +1 - thanks @lobpcg !

File: scipy/sparse/linalg/_eigen/lobpcg/tests/test_lobpcg.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rgommers) [reply scipy/sparse/linalg/_eigen/lobpcg/tests/test_lobpcg.py:433]: This sounds fine to me. These sparse tests are indeed very slow, and in quite a few cases over-parametrized.

File: scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rgommers) [reply scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:86]: If it's an arbitrary exception from within a user-defined callable, which indeed seems to be the case here, then `Exception should be okay.

File: scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rgommers) [reply scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:208]: This we cannot do, it looks like it will be a backwards compatibility break for a good fraction of users who use this.

**Review by rgommers — APPROVED** [review]
I re-ran the failed CI jobs, it all seems happy now. 

This looks like a significant improvement that closes a number of issues. Given that the 1.10.x release branch creation is planned for tomorrow, I will merge this now.

That said, there is one issue that does still need addressing. We should change the return types for for the optional outputs from `list of ndarray` to `ndarray` like that. Those outputs are easy to adjust it looks like (just use `np.split` on them right before returning them, and tweak the test for them to match). @lobpcg can you please open a follow-up PR to revert that change in output type?

**Comment by rgommers** [comment]
@lobpcg it would also be nice if you could propose a note on the improvements to `lobpcg` to add to the 1.10.0 release notes, in a comment here.

**Comment by lobpcg** [comment]
> @lobpcg it would also be nice if you could propose a note on the improvements to `lobpcg` to add to the 1.10.0 release notes, in a comment here.

@rgommers 
What does this implement/fix summary of main new features:
A simple tunable restart potentially increases the attainable accuracy for edge cases.
Internal postprocessing runs one final exact Rayleigh-Ritz method giving more accurate and orthonormal eigenvectors.
Output the computed iterate with the smallest max norm of the residual and drop the history of subsequent iterations.
Remove the check for LinearOperator format input and thus allow a simple function handle of a callable object as an input.
Fool-proof enhancements, attempting to guess and fix user errors in the input, rather than letting the execution to break.
